File: 4.py
# Backtracking Solution: N-Queens Problem

def is_safe(board, row, col, n):
    # Check column
    for i in range(row):
        if board[i][col] == 1:
            return False

    # Check left diagonal
    i, j = row-1, col-1
    while i >= 0 and j >= 0:
        if board[i][j] == 1:
            return False
        i -= 1
        j -= 1

    # Check right diagonal
    i, j = row-1, col+1
    while i >= 0 and j < n:
        if board[i][j] == 1:
            return False
        i -= 1
        j += 1

    return True


def solve_nqueens(board, row, n):
    if row == n:
        return True

    for col in range(n):
        if is_safe(board, row, col, n):
            board[row][col] = 1

            if solve_nqueens(board, row + 1, n):
                return True

            # Backtrack
            board[row][col] = 0

    return False


# ---------------- Main ----------------
n = 4
board = [[0]*n for _ in range(n)]

if solve_nqueens(board, 0, n):
    print("Solution:")
    for row in board:
        print(row)
else:
    print("No solution exists")


## output
# Solution:
# [0, 1, 0, 0]
# [0, 0, 0, 1]
# [1, 0, 0, 0]
# [0, 0, 1, 0]

File: t/4.cpp
#include <iostream>
using namespace std;

#define N 4  // Change N for different board sizes

// Function to print the solution
void printSolution(int board[N][N]) {
    cout << "Solution:\n";
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cout << board[i][j] << " ";
        }
        cout << endl;
    }
}

// Check if it's safe to place queen at board[row][col]
bool isSafe(int board[N][N], int row, int col) {

    // Check column
    for (int i = 0; i < row; i++) {
        if (board[i][col])
            return false;
    }

    // Check left diagonal
    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j])
            return false;
    }

    // Check right diagonal
    for (int i = row, j = col; i >= 0 && j < N; i--, j++) {
        if (board[i][j])
            return false;
    }

    return true;
}

// Backtracking function
bool solveNQ(int board[N][N], int row) {

    // If all queens are placed
    if (row == N)
        return true;

    // Try placing queen in all columns
    for (int col = 0; col < N; col++) {

        if (isSafe(board, row, col)) {

            // Place queen
            board[row][col] = 1;

            // Recur for next row
            if (solveNQ(board, row + 1))
                return true;

            // Backtrack
            board[row][col] = 0;
        }
    }

    return false;
}

// Main function
int main() {
    int board[N][N] = {0};

    if (solveNQ(board, 0)) {
        printSolution(board);
    } else {
        cout << "No solution exists";
    }

    return 0;
}